// miniapp-builder-shared/lib/platformMap.js
"use strict";

var _module$exports;

var _require = require('./constants'),
    MINIAPP = _require.MINIAPP,
    WECHAT_MINIPROGRAM = _require.WECHAT_MINIPROGRAM,
    BYTEDANCE_MICROAPP = _require.BYTEDANCE_MICROAPP,
    BAIDU_SMARTPROGRAM = _require.BAIDU_SMARTPROGRAM,
    KUAISHOU_MINIPROGRAM = _require.KUAISHOU_MINIPROGRAM,
    QUICKAPP = _require.QUICKAPP;

module.exports = (_module$exports = {}, _module$exports[MINIAPP] = {
  type: 'ali',
  name: 'Alibaba MiniApp',
  apiNamespace: 'my',
  nativeConfigFileName: 'mini.project.json',
  extension: {
    xml: '.axml',
    css: '.acss',
    script: '.sjs'
  }
}, _module$exports[WECHAT_MINIPROGRAM] = {
  type: 'wechat',
  name: 'WeChat MiniProgram',
  apiNamespace: 'wx',
  nativeConfigFileName: 'project.config.json',
  extension: {
    xml: '.wxml',
    css: '.wxss',
    script: '.wxs'
  }
}, _module$exports[BYTEDANCE_MICROAPP] = {
  type: 'bytedance',
  name: 'ByteDance MicroApp',
  apiNamespace: 'tt',
  nativeConfigFileName: 'project.config.json',
  extension: {
    xml: '.ttml',
    css: '.ttss',
    script: '.sjs'
  }
}, _module$exports[BAIDU_SMARTPROGRAM] = {
  type: 'baidu',
  name: 'Baidu SmartProgram',
  apiNamespace: 'swan',
  nativeConfigFileName: 'project.swan.json',
  extension: {
    xml: '.swan',
    css: '.css',
    script: '.sjs'
  }
}, _module$exports[KUAISHOU_MINIPROGRAM] = {
  type: 'kuaishou',
  name: 'KuaiShou MiniProgram',
  apiNamespace: 'ks',
  nativeConfigFileName: 'project.config.json',
  extension: {
    xml: '.ksml',
    css: '.css',
    script: '.wxs'
  }
}, _module$exports[QUICKAPP] = {
  type: 'quickapp',
  name: 'QuickApp',
  extension: {
    xml: '.ux',
    css: '.css'
  }
}, _module$exports);
