// driver-weex/es/index.js
import { setRpx, convertUnit } from 'style-unit';
var STYLE = 'style';
var ID = 'id';
var TEXT = 'text';
var CHILDREN = 'children';
var EVENT_PREFIX_REGEXP = /^on[A-Z]/;
var ARIA_PREFIX_REGEXP = /^aria-/;
var HYPHEN_REGEXP = /\-(\w)/;
var EMPTY = '';

function updateWeexTextValue(node) {
  var value = node.children.map(function (child) {
    // Comment node type
    return child.nodeType === 8 ? child.value : EMPTY;
  }).join(EMPTY);
  node.setAttr('value', value);
}

var nodeMaps = {};
/* global __weex_document__ */

var document = typeof __weex_document__ === 'object' ? __weex_document__ : typeof document === 'object' ? document : null;
export function getElementById(id) {
  return nodeMaps[id];
}
export function createBody(type, props) {
  if (document.body) {
    return document.body;
  }

  var documentElement = document.documentElement;
  var body = document.createBody(type, props);
  documentElement.appendChild(body);
  return body;
}
export function createComment(content) {
  return document.createComment(content);
}
export function createEmpty() {
  return createComment(EMPTY);
}
export function createText(text) {
  // Use comment node type mock text node
  return createComment(text);
}
export function updateText(node, text) {
  node.value = text;
  updateWeexTextValue(node.parentNode);
}
export function createElement(type, props) {
  if (props === void 0) {
    props = {};
  }

  var style = {};
  var originStyle = props.style;

  if (originStyle) {
    for (var prop in originStyle) {
      style[prop] = convertUnit(originStyle[prop], prop);
    }
  }

  var node = document.createElement(type, {
    style: style
  });

  for (var _prop in props) {
    var value = props[_prop];

    if (_prop === CHILDREN) {
      continue;
    }

    if (value != null) {
      if (_prop === STYLE) {
        continue;
      } else if (EVENT_PREFIX_REGEXP.test(_prop)) {
        var eventName = _prop.slice(2).toLowerCase();

        addEventListener(node, eventName, value, props);
      } else {
        setAttribute(node, _prop, value);
      }
    }
  }

  return node;
}
export function appendChild(node, parent) {
  parent.appendChild(node);

  if (parent.type === TEXT) {
    updateWeexTextValue(parent);
  }
}
export function removeChild(node, parent) {
  parent = parent || node.parentNode;
  var id = node.attr && node.attr[ID];

  if (id != null) {
    nodeMaps[id] = null;
  }

  parent.removeChild(node);

  if (parent.type === TEXT) {
    updateWeexTextValue(parent);
  }
}
export function replaceChild(newChild, oldChild, parent) {
  parent = parent || oldChild.parentNode;
  var previousSibling = oldChild.previousSibling;
  var nextSibling = oldChild.nextSibling;
  removeChild(oldChild, parent);

  if (previousSibling) {
    insertAfter(newChild, previousSibling, parent);
  } else if (nextSibling) {
    insertBefore(newChild, nextSibling, parent);
  } else {
    appendChild(newChild, parent);
  }
}
export function insertAfter(node, after, parent) {
  parent = parent || after.parentNode;
  parent.insertAfter(node, after);

  if (parent.type === TEXT) {
    updateWeexTextValue(parent);
  }
}
export function insertBefore(node, before, parent) {
  parent = parent || before.parentNode;
  parent.insertBefore(node, before);

  if (parent.type === TEXT) {
    updateWeexTextValue(parent);
  }
}
export function addEventListener(node, eventName, eventHandler, props) {
  // https://github.com/apache/incubator-weex/blob/master/runtime/vdom/Element.js#L421
  var params = props[eventName + 'EventParams'];
  return node.addEvent(eventName, eventHandler, params);
}
export function removeEventListener(node, eventName, eventHandler) {
  return node.removeEvent(eventName, eventHandler);
}
export function removeAttribute(node, propKey, propValue) {
  if (propKey == ID) {
    nodeMaps[propValue] = null;
  } // Weex native will crash when pass null value


  return node.setAttr(propKey, undefined, false);
}
export function setAttribute(node, propKey, propValue) {
  if (propKey == ID) {
    nodeMaps[propValue] = node;
  } // Weex only support `ariaLabel` format, convert `aria-label` format to camelcase


  if (ARIA_PREFIX_REGEXP.test(propKey)) {
    propKey = propKey.replace(HYPHEN_REGEXP, function (m, p) {
      return p.toUpperCase();
    });
  }

  return node.setAttr(propKey, propValue, false);
}
export function setStyle(node, style) {
  for (var prop in style) {
    // Translate `rpx` to weex `px`
    style[prop] = convertUnit(style[prop], prop);
  }

  node.setStyles(style);
}
export function beforeRender() {
  // Turn off batched updates
  document.open(); // Set `rpx` unit converter

  setRpx(1);
}
export function afterRender() {
  if (document.listener && document.listener.createFinish) {
    document.listener.createFinish();
  } // Turn on batched updates


  document.close();
}
