// build-plugin-app-core/lib/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var path = require("path");
var globby = require("globby");
var fs = require("fs-extra");
var generator_1 = require("./generator");
var constant_1 = require("./constant");
var dev_1 = require("./dev");
var config_1 = require("./config");
var getBuildConfig_1 = require("./utils/getBuildConfig");
// eslint-disable-next-line
var chalk = require('chalk');
// eslint-disable-next-line
var _a = require('miniapp-builder-shared').constants, MINIAPP = _a.MINIAPP, WECHAT_MINIPROGRAM = _a.WECHAT_MINIPROGRAM, BAIDU_SMARTPROGRAM = _a.BAIDU_SMARTPROGRAM, KUAISHOU_MINIPROGRAM = _a.KUAISHOU_MINIPROGRAM, QUICKAPP = _a.QUICKAPP, BYTEDANCE_MICROAPP = _a.BYTEDANCE_MICROAPP;
var miniappPlatforms = [MINIAPP, WECHAT_MINIPROGRAM, BYTEDANCE_MICROAPP, BAIDU_SMARTPROGRAM, KUAISHOU_MINIPROGRAM];
exports.default = (function (api, options) {
    var onHook = api.onHook, context = api.context, setValue = api.setValue;
    var command = context.command, commandArgs = context.commandArgs, userConfig = context.userConfig, rootDir = context.rootDir;
    var _a = userConfig.targets, targets = _a === void 0 ? ['web'] : _a;
    var framework = options.framework;
    // Set framework field
    setValue('FRAMEWORK', framework);
    var hasJsxRuntime = (function () {
        var _a;
        try {
            // auto detect of jsx runtime
            // eslint-disable-next-line
            var tsConfig = require(path.join(rootDir, 'tsconfig.json'));
            if (((_a = tsConfig === null || tsConfig === void 0 ? void 0 : tsConfig.compilerOptions) === null || _a === void 0 ? void 0 : _a.jsx) !== 'react-jsx') {
                return false;
            }
            // ensure react/jsx-runtime
            require.resolve('react/jsx-runtime');
            return true;
        }
        catch (e) {
            return false;
        }
    })();
    // Set jsx runtime value
    setValue('HAS_JSX_RUNTIME', hasJsxRuntime);
    // Check target
    checkTargets(targets);
    // Set temporary directory
    // eg: .ice or .rax
    config_1.setTempDir(api, options);
    // Set project type
    // eg: ts | js
    config_1.setProjectType(api);
    // Modify default entry to src/app
    // eg: src/app.(ts|js)
    config_1.setEntry(api, options);
    // Set alias
    config_1.setAlias(api, options);
    // register config in build.json
    config_1.setRegisterUserConfig(api);
    // register api method
    if (commandArgs.debugRuntime) {
        console.log('[deprecated] cli option --debug-runtime is deprecated, runtime file is generated as default');
    }
    var generator = initGenerator(api, __assign(__assign({}, options), { debugRuntime: userConfig.generateRuntime, hasJsxRuntime: hasJsxRuntime }));
    config_1.setRegisterMethod(api, { generator: generator });
    // add core template for framework
    var templateRoot = path.join(__dirname, './generator/templates');
    ["./app/" + framework, './common'].forEach(function (templateDir) {
        generator.addTemplateDir(path.join(templateRoot, templateDir));
    });
    // watch src folder
    if (command === 'start') {
        dev_1.default(api, { render: generator.render });
    }
    onHook("before." + command + ".run", function () { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, generator.render()];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    }); });
});
function initGenerator(api, options) {
    var getAllPlugin = api.getAllPlugin, context = api.context, log = api.log, getValue = api.getValue;
    var userConfig = context.userConfig, rootDir = context.rootDir;
    var framework = options.framework, debugRuntime = options.debugRuntime, hasJsxRuntime = options.hasJsxRuntime;
    var plugins = getAllPlugin();
    var _a = userConfig.targets, targets = _a === void 0 ? [] : _a, _b = userConfig.ssr, ssr = _b === void 0 ? false : _b;
    var isMiniapp = targets.some(function (target) { return miniappPlatforms.includes(target); });
    var targetDir = getValue(constant_1.TEMP_PATH);
    return new generator_1.default({
        rootDir: rootDir,
        targetDir: targetDir,
        defaultData: {
            framework: framework,
            isReact: framework === 'react',
            isRax: framework === 'rax',
            isMiniapp: isMiniapp,
            ssr: ssr,
            buildConfig: JSON.stringify(getBuildConfig_1.default(userConfig)),
            hasJsxRuntime: hasJsxRuntime,
            hasTabBar: hasTabBar(rootDir + "/src", framework),
        },
        log: log,
        plugins: plugins,
        debugRuntime: debugRuntime,
    });
}
function checkTargets(targets) {
    var hasError = false;
    if (Object.prototype.toString.call(targets) === '[object Object]') {
        hasError = true;
    }
    if (typeof targets === 'string') {
        hasError = true;
    }
    if (Array.isArray(targets) && !matchTargets(targets)) {
        hasError = true;
    }
    if (hasError) {
        var msg = "\n  targets must be the array type in build.json.\n\n    e.g. { \"targets\": [\"miniapp\", \"wechat-miniprogram\"] }\n\n  if you want to describes the browserslist environments for your project.\n  you should set browserslist in build.json.\n\n    e.g. { \"browserslist\": { \"chrome\": \"58\", \"ie\": 11 } }\n";
        console.log();
        console.log(chalk.red(msg));
        console.log();
        process.exit(1);
    }
}
function matchTargets(targets) {
    return targets.every(function (target) {
        return ['web', 'weex', 'kraken', MINIAPP, WECHAT_MINIPROGRAM, BYTEDANCE_MICROAPP, BAIDU_SMARTPROGRAM, KUAISHOU_MINIPROGRAM, QUICKAPP].includes(target);
    });
}
function hasTabBar(srcDir, framework) {
    if (framework === 'rax') {
        return globby.sync(['**/app.json'], { cwd: srcDir, absolute: true }).some(function (filepath) {
            var content = fs.readJSONSync(filepath);
            return content.tabBar && !content.tabBar.custom;
        });
    }
    return false;
}
