// rax-app-renderer/lib/renderer.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
/* eslint-disable */
import { render, createElement, useState, Fragment, useLayoutEffect } from 'rax';
import { createUseRouter } from 'create-use-router';
import { isWeb, isWeex, isKraken, isNode } from 'universal-env';
import UniversalDriver from 'driver-universal';
import { setInitialData } from './initialData';
import parseSearch from './parseSearch';
var useRouter = createUseRouter({ useState: useState, useLayoutEffect: useLayoutEffect });
var driver = UniversalDriver;
var AppTabBar;
function _isNullableComponent(component) {
    return !component || (Array.isArray(component) && component.length === 0);
}
function _matchInitialComponent(fullpath, routes) {
    var initialComponent = null;
    for (var i = 0, l = routes.length; i < l; i++) {
        if (fullpath === routes[i].path || (routes[i].regexp && routes[i].regexp.test(fullpath))) {
            initialComponent = routes[i].component;
            if (typeof initialComponent === 'function')
                initialComponent = initialComponent();
            break;
        }
    }
    return Promise.resolve(initialComponent);
}
function App(props) {
    var _a;
    var staticConfig = props.staticConfig, history = props.history, routes = props.routes, InitialComponent = props.InitialComponent, pageInitialProps = props.pageInitialProps;
    var PageComponent;
    if (isNode) {
        PageComponent = InitialComponent;
    }
    else {
        PageComponent = useRouter({ history: history, routes: routes, InitialComponent: InitialComponent }).component;
    }
    // Return null directly if not matched
    if (_isNullableComponent(PageComponent))
        return null;
    var pageProps = __assign({ history: history, location: history.location }, pageInitialProps);
    var tabBarProps = {
        config: staticConfig.tabBar,
        currentPageName: history.location.pathname,
        onClick: function (item) {
            history.push(item.pageName);
        }
    };
    return (createElement(Fragment, null,
        createElement(PageComponent, __assign({}, pageProps)),
        AppTabBar && ((_a = staticConfig.tabBar) === null || _a === void 0 ? void 0 : _a.items.some(function (_a) {
            var pageName = _a.pageName, path = _a.path;
            if (!pageName) {
                pageName = path;
            }
            return pageName === history.location.pathname;
        }))
            ? createElement(AppTabBar, __assign({}, tabBarProps)) : null));
}
function raxAppRenderer(options) {
    if (!options.appConfig) {
        options.appConfig = {};
    }
    var appConfig = options.appConfig, setAppConfig = options.setAppConfig;
    setAppConfig(appConfig);
    if (process.env.__IS_SERVER__)
        return;
    renderInClient(options);
}
function renderInClient(options) {
    var _a, _b;
    return __awaiter(this, void 0, void 0, function () {
        var appConfig, buildConfig, createBaseApp, emitLifeCycles, pathRedirect, staticConfig, context, _c, pathname, search, query, initialContext_1, _d, _e, runtime, appDynamicConfig, history, initialContext, routes;
        var _this = this;
        return __generator(this, function (_f) {
            switch (_f.label) {
                case 0:
                    appConfig = options.appConfig, buildConfig = options.buildConfig, createBaseApp = options.createBaseApp, emitLifeCycles = options.emitLifeCycles, pathRedirect = options.pathRedirect, staticConfig = options.staticConfig;
                    context = {};
                    if (!((_a = window) === null || _a === void 0 ? void 0 : _a.__INITIAL_DATA__)) return [3 /*break*/, 1];
                    context.initialData = window.__INITIAL_DATA__.initialData;
                    context.pageInitialProps = window.__INITIAL_DATA__.pageInitialProps;
                    return [3 /*break*/, 3];
                case 1:
                    if (!(isWeb && ((_b = appConfig === null || appConfig === void 0 ? void 0 : appConfig.app) === null || _b === void 0 ? void 0 : _b.getInitialData))) return [3 /*break*/, 3];
                    _c = window.location, pathname = _c.pathname, search = _c.search;
                    query = parseSearch(search);
                    initialContext_1 = {
                        pathname: pathname,
                        query: query,
                    };
                    _d = context;
                    return [4 /*yield*/, appConfig.app.getInitialData(initialContext_1)];
                case 2:
                    _d.initialData = _f.sent();
                    _f.label = 3;
                case 3:
                    _e = createBaseApp(appConfig, buildConfig, context), runtime = _e.runtime, appDynamicConfig = _e.appConfig, history = _e.history;
                    setInitialData(context.initialData);
                    initialContext = {
                        pathname: '',
                        query: {},
                    };
                    // Set custom driver
                    if (typeof staticConfig.driver !== 'undefined') {
                        driver = staticConfig.driver;
                    }
                    routes = staticConfig.routes;
                    // Like https://xxx.com?_path=/page1, use `_path` to jump to a specific route.
                    pathRedirect(history, routes);
                    return [2 /*return*/, _matchInitialComponent(history.location.pathname, routes)
                            .then(function (InitialComponent) { return __awaiter(_this, void 0, void 0, function () {
                            var initialComponent, _a, props, _b, app, rootId, appInstance, rootEl, webConfig;
                            return __generator(this, function (_c) {
                                switch (_c.label) {
                                    case 0:
                                        initialComponent = InitialComponent();
                                        if (!(!context.pageInitialProps && initialComponent.getInitialProps)) return [3 /*break*/, 2];
                                        _a = context;
                                        return [4 /*yield*/, initialComponent.getInitialProps(initialContext)];
                                    case 1:
                                        _a.pageInitialProps = _c.sent();
                                        _c.label = 2;
                                    case 2:
                                        props = {
                                            staticConfig: staticConfig,
                                            history: history,
                                            routes: routes,
                                            InitialComponent: InitialComponent,
                                            pageInitialProps: context.pageInitialProps
                                        };
                                        _b = appDynamicConfig.app, app = _b === void 0 ? {} : _b;
                                        rootId = app.rootId;
                                        appInstance = getRenderAppInstance(runtime, props, options);
                                        ;
                                        // Emit app launch cycle
                                        emitLifeCycles();
                                        rootEl = isWeex || isKraken ? null : document.getElementById(rootId);
                                        if (isWeb && rootId === null)
                                            console.warn('Error: Can not find #root element, please check which exists in DOM.');
                                        webConfig = buildConfig.web || {};
                                        return [2 /*return*/, render(appInstance, rootEl, { driver: driver, hydrate: webConfig.hydrate || webConfig.snapshot || webConfig.ssr })];
                                }
                            });
                        }); })];
            }
        });
    });
}
export function getRenderAppInstance(runtime, props, options) {
    var _a;
    var ErrorBoundary = options.ErrorBoundary, TabBar = options.TabBar, _b = options.appConfig, appConfig = _b === void 0 ? {} : _b;
    var _c = appConfig.app || {}, ErrorBoundaryFallback = _c.ErrorBoundaryFallback, onErrorBoundaryHander = _c.onErrorBoundaryHander, onErrorBoundaryHandler = _c.onErrorBoundaryHandler, errorBoundary = _c.errorBoundary;
    AppTabBar = TabBar;
    var AppProvider = (_a = runtime === null || runtime === void 0 ? void 0 : runtime.composeAppProvider) === null || _a === void 0 ? void 0 : _a.call(runtime);
    var RootComponent = function () {
        if (AppProvider) {
            return (createElement(AppProvider, null,
                createElement(App, __assign({}, props))));
        }
        return createElement(App, __assign({}, props));
    };
    var Root = createElement(RootComponent, null);
    if (process.env.NODE_ENV === 'development') {
        if (onErrorBoundaryHandler) {
            console.error('Please use onErrorBoundaryHandler instead of onErrorBoundaryHander');
        }
    }
    if (errorBoundary && ErrorBoundary) {
        return (createElement(ErrorBoundary, { Fallback: ErrorBoundaryFallback, onError: onErrorBoundaryHandler || onErrorBoundaryHander }, Root));
    }
    else {
        return Root;
    }
}
export default raxAppRenderer;
