// build-plugin-ice-store/lib/utils/getPath.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getRaxPagesPath = exports.getPageModelPath = exports.getPageStorePath = exports.getPagePath = exports.getAppModelsPath = exports.getAppStorePath = void 0;
var path = require("path");
var fse = require("fs-extra");
function getAppStorePath(_a) {
    var rootDir = _a.rootDir, srcDir = _a.srcDir, projectType = _a.projectType;
    // e.g: src/store.ts
    var appStoreFilePath = path.join(rootDir, srcDir, "store." + projectType);
    return appStoreFilePath;
}
exports.getAppStorePath = getAppStorePath;
function getAppModelsPath(_a) {
    var rootDir = _a.rootDir, srcDir = _a.srcDir;
    // e.g: src/models/*
    return path.join(rootDir, srcDir, 'models');
}
exports.getAppModelsPath = getAppModelsPath;
/**
 * return absolute page path. e.g.: /project/src/pages/Home
 */
function getPagePath(_a) {
    var rootDir = _a.rootDir, srcDir = _a.srcDir, pageName = _a.pageName;
    return path.join(rootDir, srcDir, 'pages', pageName);
}
exports.getPagePath = getPagePath;
function getPageStorePath(_a) {
    var rootDir = _a.rootDir, srcDir = _a.srcDir, projectType = _a.projectType, pageName = _a.pageName;
    var pageNameDir = getPagePath({ rootDir: rootDir, srcDir: srcDir, pageName: pageName });
    // e.g: src/pages/${pageName}/store.ts
    var pageStoreFilePath = path.join(pageNameDir, "store." + projectType);
    return pageStoreFilePath;
}
exports.getPageStorePath = getPageStorePath;
function getPageModelPath(_a) {
    var rootDir = _a.rootDir, srcDir = _a.srcDir, projectType = _a.projectType, pageName = _a.pageName;
    var pageNameDir = getPagePath({ rootDir: rootDir, srcDir: srcDir, pageName: pageName });
    // e.g: src/pages/${pageName}/models/*
    var pageModelsDir = path.join(pageNameDir, 'models');
    // e.g: src/pages/${pageName}/model.ts
    var pageModelFile = path.join(pageNameDir, "model." + projectType);
    return { pageNameDir: pageNameDir, pageModelsDir: pageModelsDir, pageModelFile: pageModelFile };
}
exports.getPageModelPath = getPageModelPath;
function getRaxPagesPath(rootDir) {
    var absoluteAppJSONPath = path.join(rootDir, 'src/app.json');
    var appJSON = fse.readJSONSync(absoluteAppJSONPath);
    var routes = appJSON.routes;
    var pagesPath = routes.map(function (route) { return route.source; });
    return pagesPath;
}
exports.getRaxPagesPath = getRaxPagesPath;
